lib/drivers/docker/index.js

```diff
@@ -0,0 +1,294 @@
+  }
+};
+
+DockerDriver.prototype.removeInstance = function(id, cb) {
+  debug('DockerDriver.removeInstance(%j)', id);
+  var instance = this._instance(id);
+  if (instance.current) {
+    instance.current.kill('SIGKILL', cb);
+    delete instance.current;
+  } else {
+    setImmediate(cb);
+  }
+};
+
+DockerDriver.prototype.deployInstance = function(id, req, res) {
+  debug('DockerDriver.deployInstance(%j)', id);
+  var self = this;
+  var instance = this._containerById(id);
+  var image = Image.from(this, id, this.baseDir, req, res);
+  instance.next = image;
+  image.on('error', function(err) {
+    console.error('error deploying: %s, %j', err, err);
+    throw err;
+  });
+  image.on('image', function() {
+    debug('Deployment image:', image.name);
+    self.instances[id].next.image = image;
+    self.startInstance(id, function(err) {
+      debug('deployInstance -> startInstance -> ', err);
+    });
+  });
+};
+
+DockerDriver.prototype.startInstance = function(id, cb) {
+  debug('DockerDriver.startInstance(%j, %j)', id, cb.name);
+  var instance = this._instance(id);
+  var toLaunch = instance.next || instance.currentImage || instance.previous;
+  var img = toLaunch.image;
+  var self = this;
+  this.server.getInstanceEnv(id, function(err, env) {
+    if (err) return cb(err);
+    // ensure environment variable values have been stringified
+    env = _.mapValues(env, String);
+    var opts = _.merge(instance.startOpts, {env: env});
+    instance.log.enableGC();
+    var previous = instance.current;
+    instance.current = img.start(instance.log, opts);
+    instance.previous = instance.currentImage;
+    instance.currentImage = toLaunch;
+    delete instance.next;
+    instance.current.on('request', function(req) {
+      debug('bubbling up request from instance %j:', id, req);
+      self.emit('request', id, req);
+    });
+    instance.current.on('error', function(err) {
+      console.error('error from docker child %j:', id, err);
+    });
+    instance.current.on('starting', function() {
+      instance.ports = [];
+      if (previous) {
+        previous.kill('SIGKILL', function() {
+          previous.destroy();
+        });
+      }
+    });
+    instance.current.on('exit', function(err) {
+      if (isError(err) && err.statusCode === 404) {
+        self.startInstance(id, function() {
+          console.log('restarted instance after container disappeared (exit)');
+        });
+      }
+    });
+    if (cb) {
+      instance.current.on('connected', cb);
+    }
+  });
+};
+
+DockerDriver.prototype._onRequest = function(instanceId, req) {
+  debug('DockerDriver._onRequest(%j)', instanceId, req);
+  var instance = this._instance(instanceId);
+  switch (req.cmd) {
+    case 'listening':
+      // emit a listening event each time a _new_ port is listened on
+      if (instance.ports.indexOf(req.address.port) < 0) {
+        instance.ports.push(req.address.port);
+        this.emit('listening', instanceId, req.address);
```

Comment:
is the only purpose of `instance.ports` to avoid duplicate emit of the listening event? I note that an app that listens on ipv6 an ipv4 on same port, only one of the listening events will get through. Is there harm in allowing duplicate addresses? I'd hope not.

Also, I think if .port is null (because it's a unix domain socket), that this will go awry.